### PR TITLE
[DO NOT MERGE] [WIP] StringGuts

### PR DESCRIPTION
**Pull request by milseman**
An initial implementation of a more resilient 2-word representation for String. It uses BridgeObject (for now) to achieve a high degree of resilience. In the future, we'll add more small string forms, establish UnsafeString, and generally rejigger String's views to be properly layered on top.

Notes:

* Self slices are currently implemented by forming `_NSContiguousStrings` to hold the sliced up _StringCore. This incurs overhead. Long term, we could support owned self slices, but for now it would be fruitful to try to eliminate all uses of self-slicing _StringCore from the stdlib. The most common occurrence (but by no means the only) is through _ephemeralString. Instead, for anything that can't possibly escape, we should convert to an UnsafeString, guarantee the lifetime, and slice that up.

* We cannot utilize small representations effectively until we start to wean ourselves off of `_LegacyStringCore`. Ideally (depending on project planning needs) we would be fully weaned off before adding small representations. We'd also like to get close to performance parity where it matters prior, to better gauge the impact of such optimizations.

* The spiritual predecessor to this effort is at https://github.com/dabrahams/swift/commits/string-recore, and while the physical implementation differs, that's a great demonstration of what we're trying to accomplish.

### CODE REVIEW

**Comment by milseman** [comment]
Note, some tests are testing _StringCore's ABI, which is obviously changed here. They will be failing until updated.

**Comment by milseman** [comment]
@swift-ci please benchmark

**Comment by milseman** [comment]
@swift-ci please benchmark

**Comment by dabrahams** [comment]
It seems like a shame to throw out all this work, which performs well since Arnold's fixes went in: https://github.com/dabrahams/swift/commits/string-recore

**Comment by milseman** [comment]
This is inspired by that work (in fact one of the commits explicitly linked to it, but I think that got dropped during a recent squashing). More and more of it will be moved over, as needed. The work can't be taken verbatim, as it's untenable for ABI stability, at least as the Swift platform exists today.

As I'm sure you're well aware, we don't want the project to continuously succumb to the sunk cost fallacy. Not all code lives forever. Do you have any specific, concrete concerns here?

**Comment by dabrahams** [comment]
I guess my main concern was the use of BridgeObject instead of an enum, and I confess I don't understand why there would be any special resilience issues here.

**Comment by milseman** [comment]
Ah, yes. `BridgeObject` gives us precise control over the bit representation. This is useful for resilience because it affords us a means of relaxing resilience bit-by-bit to an extent that wouldn't be offered through an enum. At least, without significant contortions.

An enum might make sense in the future (perhaps for other types, as we hopefully will be "done" with String at some point). The issue today is that enum layout is not settled, and I need to break a dependency between optimal enum layout and the new String.

While this *is* software, and in software all decidable things are possible, I'm trying to carefully weigh the ugliness of `BridgeObject` over the benefits of being decoupled from parts of the system that are still in flux. `BridgeObject` gives us the following kind of pseudo-structure with a smaller "attack surface" of potential breakage:

```swift
enum _StringGuts {
  case ManagedNative(UInt6, NativeObjectAddressBits, UInt3, UInt64)
  case ManagedCocoa(UInt6, ForeignObjectAddressBits, UInt3, UInt64)
  case Value(UInt127)
}
```

Of course, `UInt6`, `UInt127`, `NativeObjectAddressBits`, `ForeignObject`, `ForeignObjectAddressBits`, `UInt3`, `UInt127` all don't exist. Enum layout as it currently exists fails to find and use spare bits as it needs to. Even though this enum has many extra inhabitants, enum layout as it currently exists won't use them for Optional. Any progress made on enum layout in Swift 4.x or Swift 5 should be focused on those enums that must be ABI stable in that timeframe (highest priority), while optimal enum layout for very complicated cases like this can come later, worst case.

**Comment by milseman** [comment]
@swift-ci please benchmark

**Comment by milseman** [comment]
Argh.

**Comment by milseman** [comment]
@swift-ci please smoke benchmark

**Comment by dabrahams** [comment]
What are “self“ slices, and why are they interesting?  Is there a reason a wrapper over RangeReplaceableMutableBidirectionalSlice<String> is insufficient as the slice of a String?

**Comment by milseman** [comment]
@dabrahams, I lost my train of thought and why I was being careful to not reject them out-of-hand. I don't think owned self slices are very interesting and would prefer use of one of the slice types. Unowned self slices are far more useful and interesting: a lexically scoped analogue of `ephemeralString`, forming Strings around immortal storage, etc. Such a construct can also enable optimizations in conjunction with more ownership annotations/tracking in the language's future evolution. What I'm mostly referring to in this PR is an artifact of interop with the old `_StringCore`, whose representation supports them.

But, this is software and anything's possible with enough shenanigans. The argument for owned self slices would be so that String-as-common-currency-type can represent a Substring. This is probably a bad idea, for reasons debated in the one-type vs two-type arguments last year.

**Comment by milseman** [comment]
@swift-ci please smoke benchmark

**Comment by dabrahams** [comment]
I'm afraid I don't know how to read that because you didn't answer my first question: What are “self“ slices?

**Comment by milseman** [comment]
Aha, that is a confusing term. What I mean by "self slice" is a `String` that's really a slice. That is, its `_StringCore`, which is capable of being its own slice type, is being used as its own slice type.

**Comment by milseman** [comment]
@swift-ci please smoke benchmark

**Comment by milseman** [comment]
@swift-ci please smoke benchmark

**Comment by milseman** [comment]
@swift-ci please smoke benchmark

**Comment by milseman** [comment]
@swift-ci please smoke benchmark

**Comment by milseman** [comment]
@swift-ci please smoke benchmark

**Comment by milseman** [comment]
@swift-ci please smoke benchmark

**Comment by milseman** [comment]
@swift-ci please smoke benchmark

**Comment by milseman** [comment]
@swift-ci please smoke benchmark

**Comment by milseman** [comment]
@swift-ci please smoke benchmark

**Comment by milseman** [comment]
@swift-ci please smoke benchmark

**Comment by milseman** [comment]
@swift-ci please smoke benchmark

**Comment by milseman** [comment]
The above was checking if, in addition to storing the count at a known place, we also store the pointer, making things like `_unmanagedContiguous` near-trivial.

**Comment by milseman** [comment]
@swift-ci please smoke benchmark

**Comment by milseman** [comment]
@swift-ci please smoke benchmark

File: stdlib/public/core/StringGuts.swift [hunk 1]
```diff
@@ -0,0 +1,1571 @@
...
+    _sanityCheck(CodeUnit.bitWidth == self.bitWidth)
+    return _bridgeObject(toNative: _object) as! _SwiftStringStorage<CodeUnit>
+  }
+
+  @_inlineable
+  public // @testable
+  var _nativeRawStorage: _SwiftRawStringStorage {
+    @inline(__always) get {
+      _sanityCheck(_isNative)
+      return _bridgeObject(toNative: _object) as! _SwiftRawStringStorage
```
Comment (milseman) [inline stdlib/public/core/StringGuts.swift]: @lorentey, do you need to unflag it (in case it's UTF16)?

File: test/stdlib/NewString.swift [hunk 2]
```diff
@@ -14,45 +14,33 @@ func hexAddrVal<T>(_ x: T) -> String {
...
+  if x._isNative {
+    return "Native("
+      + "owner: \(hexAddrVal(x._owner)), "
+      + "count: \(x.count), "
+      + "capacity: \(x.capacity))"
+  } else if x._isNonTaggedCocoa {
+    return "Cocoa("
+      + "owner: \(hexAddrVal(x._owner)), "
+      + "count: \(x.count))"
+  } else if x._isTaggedCocoa {
```
Comment (milseman) [inline test/stdlib/NewString.swift]: @lorentey I believe the structure changed around here (i.e. there is no more _isTaggedCocoa var).

**Comment by milseman** [comment]
@swift-ci please smoke benchmark

**Comment by milseman** [comment]
Err, last smoke benchmark failed with
```
/Users/buildnode/jenkins/workspace/swift-PR-osx-smoke-perf/swift/lib/IRGen/IRGenDebugInfo.cpp:1981:47: error: no member named 'FindAllocaDbgDeclare' in namespace 'llvm'
  if (isa<llvm::AllocaInst>(Storage) && llvm::FindAllocaDbgDeclare(Storage))
                                        ~~~~~~^
1 error generated.
```

I think that might be us lagging behind master too much. I have a branch that's rebased from master (at least, master as of Friday), so I'll see if that works if this one doesn't.

**Comment by milseman** [comment]
That was a rebase from Friday. I'll need to do another one. Hopefully that's enough to get the benchmarks running, though.

**Comment by milseman** [comment]
@swift-ci please smoke benchmark

**Comment by milseman** [comment]
@swift-ci please smoke benchmark

**Comment by milseman** [comment]
@shahmishal is something wrong with the smoke benchmark bots, or is the build failing? Looking at the console I don't see any links to a failed build

**Comment by milseman** [comment]
@swift-ci please smoke benchmark

**Comment by milseman** [comment]
@swift-ci please test

**Comment by milseman** [comment]
@swift-ci please smoke benchmark

**Comment by milseman** [comment]
@swift-ci please test

**Comment by milseman** [comment]
@swift-ci please smoke benchmark

File: stdlib/public/core/StringGuts.swift [hunk 3]
```diff
@@ -0,0 +1,1653 @@
...
+    case (false, false):
+      return _StringGuts(
+        _copyToNativeStorage(of: UTF16.CodeUnit.self, from: range))
+    }
+  }
+
+  @_inlineable
+  @inline(__always)
+  public // @testable
+  mutating func isUniqueNative() -> Bool {
```
Comment (milseman) [inline stdlib/public/core/StringGuts.swift]: @lorentey. BTW, is this safe and sufficient in a multi-threaded environment? That is, do we instead need to ensure uniqueness for the entire execution of a method, or is that guaranteed / implied by the fact that self itself is local, and any copies would invariably bump the count? I don't know the rules here. CC @aschwaighofer @atrick @moiseev

Comment (aschwaighofer) [reply stdlib/public/core/StringGuts.swift]: The standard library does not need to guard against language level races. A language level race is just that and needs to be addressed by the user of the standard library.

For example: 

‘’’
var a: String  // global

Thread 1:

a.mutate()

Thread 2:

a.mutate() // race on ‘a’
‘’’

Will race without any additional synchronization in the user program.

**Comment by milseman** [comment]
@swift-ci please smoke benchmark

**Comment by milseman** [comment]
@lorentey I want to see how that run goes. I have just rebased from master again, and after I get those results (and assuming my tests pass) I'll push that rebase.

File: stdlib/public/core/StringGuts.swift [hunk 4]
```diff
@@ -0,0 +1,1355 @@
...
+
+    // Note that we have to perform this operation here, and not as a (even
+    // mutating) method on our _StringObject to avoid all chances of a semantic
+    // copy.
+    //
+    // FIXME: Super hacky. Is there a better way?
+    defer { _fixLifetime(self) }
+    var bitPattern = _object.payloadBits
+    return _isUnique_native(&bitPattern)
+  }
```
Comment (milseman) [inline stdlib/public/core/StringGuts.swift]: CC @atrick, this is what we were discussing earlier.

**Comment by milseman** [comment]
@swift-ci please smoke benchmark

**Comment by milseman** [comment]
@swift-ci please smoke benchmark

**Comment by milseman** [comment]
@swift-ci please smoke benchmark

**Comment by milseman** [comment]
That had a 2x+ impact on StringBuilderLong, which should now magnify the constant-costs we're facing.

@lorentey, I noticed that append ends up loading type metadata even if specialized. My guess is it's the bit width checks inside UnmanagedString._copy. Have you seen issues here?

Also, the String.swift validation test just started failing for me. It's expecting String range-based subscript to do bounds checking, and currently it doesn't. I don't see how this has changed recently, so it's a little confusing for me.

**Comment by milseman** [comment]
@swift-ci please smoke benchmark

**Comment by milseman** [comment]
@swift-ci please smoke benchmark

**Comment by lorentey** [comment]
> @lorentey, I noticed that append ends up loading type metadata even if specialized. My guess is it's the bit width checks inside UnmanagedString._copy. Have you seen issues here?
>
> Edit: Actually, what are your plans for UnmanagedString being generic? I'm seeing all kinds of witness table accesses for UInt8 in UmanagedString's measureExtendedGraphemeClusterForward for StringMatch. They may not contribute much in total, but it's probably going to limit the optimizer in other areas as well.

I really want to split `_UnmanagedString` into two non-generic types; the `CodeUnit` tests are not only slow sometimes, but they also look clumsy. `_StringVariant` will reduce code duplication, but we can also add a new internal protocol for operations that are common between the two contiguous views. (There is definitely no need to do any gybbing.)

> Also, the String.swift validation test just started failing for me. It's expecting String range-based subscript to do bounds checking, and currently it doesn't. I don't see how this has changed recently, so it's a little confusing for me.

I'm not quite sure why the test would fail only now, but some of the bounds checking was lost during the migration to unmanaged string views. My new PR restores most of them, but we should also do an audit pass for any checks that are still missing (especially on API that isn't currently included in the test suite). For now, I added methods to `_StringVariant` for explicit boundary checking; I think a better option would be to make bounds checking the default and explicitly opt-in to unchecked access via `foo[unchecked: i]` subscripts -- this can be done as part of the audit.

**Comment by milseman** [comment]
@swift-ci please smoke benchmark

**Comment by milseman** [comment]
@swift-ci please smoke benchmark

**Comment by milseman** [comment]
@swift-ci please smoke benchmark

**Comment by milseman** [comment]
@swift-ci please smoke benchmark

File: stdlib/public/core/StringGuts.swift (hunk 4 above)
Comment (atrick) [reply stdlib/public/core/StringGuts.swift]: I have a better idea now how SILGen could be fixed to allow isUnique(&self.property). I'm not sure when someone would get around to fixing it, but filing a bug for any case you come across would be reasonable and actionable.

**Comment by milseman** [comment]
Rebased

@swift-ci please smoke test

**Comment by milseman** [comment]
Scratch that, let's see how bad it is:

@swift-ci please test

**Comment by milseman** [comment]
@swift-ci please test

**Comment by milseman** [comment]
@swift-ci please test

**Comment by milseman** [comment]
@swift-ci please smoke benchmark

**Comment by milseman** [comment]
@swift-ci please test

**Comment by milseman** [comment]
@swift-ci please smoke benchmark

**Comment by milseman** [comment]
@swift-ci please test Linux Platform

**Comment by milseman** [comment]
(MacOS tests still chugging away at https://ci.swift.org/job/swift-PR-osx/2462/console)

**Comment by milseman** [comment]
Failures:
```
    Swift(iphonesimulator-i386) :: stdlib/Unicode.swift.gyb
    Swift(iphonesimulator-i386) :: stdlib/TestUUID.swift
```

Weird, when I run locally I don't get those. @lorentey, if you get a chance, could you investigate?

**Comment by milseman** [comment]
@swift-ci please test linux platform

**Comment by milseman** [comment]
I don't understand how Linux is failing with:
```
error: initializer for class '_SwiftRawStringStorage' is '@_inlineable' and must delegate to another initializer
```
Even after https://github.com/apple/swift/pull/12442/commits/f932bef39670fb2436e2d0de818582a3031cb941. MacOS builds just fine, there's just 2 test failures currently.

**Comment by milseman** [comment]
@swift-ci please test os x platform

**Comment by milseman** [comment]
@swift-ci please test

**Comment by milseman** [comment]
(still trying to reproduce the iphonesimulator failure, which works locally for me, I'm trying the bot's exact configuration now)

**Comment by milseman** [comment]
@lorentey, the only remaining failure is on Linux, it seems that a bounds check is not happening: `ForeignIndexes/subscript(Range)/OutOfBoundsTrap/1`. Were you able to get a full validation run on Linux clean, and do you recall if there was any issue here?

https://ci.swift.org/job/swift-PR-Linux/2490/testReport/junit/Swift(linux-x86_64)/stdlib/String_swift/

**Comment by lorentey** [comment]
There were no issues left in my local validation test runs -- I ran them on Linux/x86_64, macOS/x86_64 and watchOS-simulator/i386 for [#18](https://github.com/milseman/swift/pull/18).

I just finished a compiler build on Linux from the latest sources. Subscripting with out-of-bounds indices correctly triggers the precondition on [StringVariant.swift:90](https://github.com/milseman/swift/blob/string_guts/stdlib/public/core/StringVariant.swift#L90) and `build-script -rT` reports no issues (after removing a hang on a compiler crasher test). I'm looking into what's different in the buildbot_linux environment.

**Comment by milseman** [comment]
@swift-ci please test Linux Platform

**Comment by lorentey** [comment]
@swift-ci please test Linux Platform

**Comment by milseman** [comment]
@swift-ci please test

**Comment by milseman** [comment]
@swift-ci please smoke benchmark

**Comment by milseman** [comment]
Err, MacOS was just passing.

@swift-ci please test os x platform.

**Comment by milseman** [comment]
This all passes locally for me with ./utils/build-script -R -T -i. I'll also try a --no-assertions...

@swift-ci please test OS X platform

**Comment by milseman** [comment]
@swift-ci please test os x platform

**Comment by milseman** [comment]
@swift-ci please test

**Comment by milseman** [comment]
@swift-ci please smoke benchmark

**Comment by milseman** [comment]
Failures:
```
    Swift(iphonesimulator-i386) :: stdlib/AnyHashable.swift.gyb
    Swift(iphonesimulator-i386) :: Interpreter/SDK/UIActionSheetHack.swift
    Swift(iphonesimulator-i386) :: stdlib/CodableTests.swift
    Swift(iphonesimulator-i386) :: stdlib/TestJSONEncoder.swift
```

**Comment by milseman** [comment]
So I built locally identically to the bot, and these don't consistently reproduce.

@swift-ci please clean test os x platform

**Comment by milseman** [comment]
Rebased, re-built locally, still cannot repro...

@swift-ci please test os x platform

File: stdlib/public/SDK/Foundation/NSString.swift [hunk 5]
```diff
@@ -23,8 +23,8 @@ public class NSSimpleCString {}
 public class NSConstantString {}
 
 // Called by the SwiftObject implementation.
-public func _convertStringToNSString(_ string: String) -> NSString {
-  return string._bridgeToObjectiveC()
+public func _getDescription<T>(_ x: T) -> NSString {
+  return String(reflecting: x)._bridgeToObjectiveC()
```
[the remark this answers is not in the document]
Comment (lorentey) [reply stdlib/public/SDK/Foundation/NSString.swift]: @milseman Oops, sorry, I missed this question earlier. This change merges `_getSummary` in the stdlib with `_convertStringToNSString` in Foundation to eliminate the need to hold a `String` value in the runtime.

`_getSummary` itself used `String(reflecting:)`, I just moved it over here.

**Comment by milseman** [comment]
(Tiny bit of a shot in the dark, but I suspect it's 32-bit pointers in the latter half of the address space).

@swift-ci please test

**Comment by milseman** [comment]
Got past iphonesimulator-i386!
🎉

**Comment by milseman** [comment]
Woo! Going to push some squashes.

**Comment by milseman** [comment]
Merge will take place in https://github.com/apple/swift/pull/14046

**Pull request closed without merging**